openquake/hazardlib/gsim/utils_chapman_guo_2021.py

```diff
@@ -0,0 +1,203 @@
+            'SA(0.75)',
+            'SA(1.0)',
+            'SA(1.5)',
+            'SA(2.0)',
+            'SA(3.0)',
+            'SA(4.0)',
+            'SA(5.0)',
+            'SA(7.5)',
+            'SA(10.0)']
+
+
+def get_psa_df(psa_df, imt):
+    """
+    Get the subset of the PSA ratio DataFrame for the given imt 
+    """
+    cols = ['zsed', 'magnitude', 'distance', f'psa_ratio_{imt}']
+    return psa_df[cols]
+
+
+def get_zscale(z_sed):
+    """
+    Provide the depth scaling factor for application of reference site
+    scaling.
+
+    :param z_sed: Depth to sediment site parameter considered in the
+                  Chapman and Guo (2021) Coastal Plains site amplification
+                  model as part of the 2023 Conterminous US NSHMP.
+    """
+    Z_CUT = 2.
+    s = 1. - np.exp((-1 * z_sed) / Z_CUT)
+    return s ** 4
+
+
+def get_fcpa(ctx, imt, z_scale, psa_df):
+    """
+    Get f_cpa param for the given sediment depth, Mw and rjb.
+
+    This function returns both f_cpa and z_scale within a dictionary which is
+    passed into the nga_east functions for computation of mean ground-motions.
+    """
+    # Set default f_cpa of zero for each site
+    f_cpa = np.full(len(ctx.vs30), 0.)
+
+    # Get sites with sed. depth scaling greater than 0
+    mask_z = z_scale > 0.
+
+    # For these sites recompute f_cpa parameter
+    if np.any(mask_z):
+        f_cpa[mask_z] = get_psa_ratio(ctx, imt, psa_df)
+        
+    # Put Coastal Plain params into a dict for passing into nga_east functions
+    coastal = {'f_cpa': f_cpa, 'z_scale': z_scale}
+
+    return coastal
+
+
+def get_fraction(lo, hi, value):
+    """
+    Get fraction of admitted value between lower and upper values.
+    """
+    return np.clip((value - lo) / (hi - lo), 0.0, 1.0)
+
+
+def get_data(psa_df, imt):
+    """
+    Get the z_sed for each z_sed, mag and rrup combination within an ndarray.
+    """
+    # Append columns with given imt
+    cols = ['zsed', 'magnitude', 'distance']
+    
+    # Make multi-idx
+    idx = pd.MultiIndex.from_product(
+        [Z, M, R], names=cols)
+
+    # Set df idx to match multi-idx
+    psa_df.set_index(cols, inplace=True)
+
+    # Align the df with multi-idx to match the rows
+    psa_df_aligned = psa_df.reindex(idx)
+
+    # Get PSA ratios into ndarray
+    data = psa_df_aligned[f'psa_ratio_{imt}'].values.reshape(
+        len(Z), len(M), len(R))
+
+    return data
+
+
+def interpolate(lo, hi, fraction):
+    """
+    Perform weighted interpolation between the lower and upper values.
+    """
+    return lo + fraction * (hi - lo)
+
+
+def get_psa_ratio(ctx, imt, psa_df):
+    """
+    Get the PSA ratio for each ctx's sediment depth, M and rrup for the
+    given IMT.
+    """
+    # Get psa data into ndarray
+    data = get_data(psa_df, imt)
+
+    # Get values per ctx into arrays
+    z = np.array([cx.z_sed for cx in ctx])
+    m = np.array([cx.mag for cx in ctx])
+    r = np.array([cx.rrup for cx in ctx])
```

Comment:
The lines above should be replaced with
```
    z = ctx.z_sed
    m = ctx.mag
    r = ctx.rrup
```
which is the way contexts are meant to be used.